gosh/framework.sh.tpl

```diff
@@ -0,0 +1,106 @@
+#!/bin/sh
+
+set -u
+
+GOSH_FAILED=0
+
+if command -v gt >/dev/null 2>&1; then
+    readonly GT_BIN="gt"
+else
+    echo "Error: gt command not found in PATH" >&2
+    exit 1
+fi
+
+# Send a signal to the test runner
+# Args:
+#   $1 - The operation to perform (e.g., start, stop, status)
+#   $2 - The name of the test
+#   $3 - The exit code of the test (optional, default is 0)
+#   $4 - An optional message to include (default is "")
+_gosh_signal() {
+    local op="${1}"
+    local test_name="${2}"
+    local exit_code="${3:-0}"
+    local message="${4:-}"
+
+    "${GT_BIN}" ctrl \
+        --op "$op" \
+        --test-name "$test_name" \
+        --exit-code "$exit_code" \
+        --addr "${GOSH_CONTROL_ADDR}" \
+        --message "${message}"
+}
+
+# Check if a function exists
+_gosh_function_exists() {
+    type "$1" 2>/dev/null | grep -q 'function\|shell function'
```

Comment:
this buys you a dependency on grep. (or busybox to provide it). Its also kind of silly to look for 'functon' or 'shell function' as 100% of the things that match the former will match the later, right?

I think its worth avoiding that here. You could use `case` like:

```
local out=""
out=$(type "$1" 2>/dev/null) || { echo "unexpected error 'type $1': $?"; return 1; }
case "$out" in
  *\ function) return 0;;
esac
return 1
}
```

Reply:
Another thing you could do here would be to just avoid the check entirely by having the template define the function, so that there always _was_  function by that name, and the test only possibly replaces it.

Reply:
nice find! I went with your suggestion instead of `grep`